environment.py

```diff
@@ -85,59 +113,48 @@ def findCellNeighbors(self):
-                        cellsInRange.append({"cell": self.grid[reflectedX][deltaY], "distance": euclideanDistance})
-                        cellsInRange.append({"cell": self.grid[reflectedX][reflectedY], "distance": euclideanDistance})        
+    def findCellRanges(self):
+        config = self.sugarscape.configuration
+        # Determine maximum range to memoize based on the maximum possible agent vision and movement from bonuses
+        maxVision = config["startingDiseases"] * max(config["diseaseVisionPenalty"][1], 0) + config["agentVision"][1]
+        maxMovement = config["startingDiseases"] * max(config["diseaseMovementPenalty"][1], 0) + config["agentMovement"][1]
+        maxAgentRange = max(maxVision, maxMovement)
+        maxDeltaX = self.width - 1 if maxAgentRange > self.width - 1 else maxAgentRange
+        maxDeltaY = self.height - 1 if maxAgentRange > self.height - 1 else maxAgentRange
+        maxDeltaRadius = max(maxDeltaX, maxDeltaY)
+        cellCoords = [(x, y) for x in range(self.width) for y in range(self.height)]
+        # Initialize ranges with all possible values
+        for x, y in cellCoords:
+            self.grid[x][y].ranges = {gridRange: {} for gridRange in range(1, maxDeltaRadius + 1)}
+
+        if config["agentVisionMode"] == "radial" and config["agentMovementMode"] == "radial":
+            self.findRadialCellRanges(maxDeltaX, maxDeltaY, maxDeltaRadius, cellCoords)
         else:
-            cellsInRange = []
-            # Iterate through the bounding box of the circle
-            for deltaX in range(max(0, startX - gridRange), min(self.width, startX + gridRange + 1)):
-                for deltaY in range(max(0, startY - gridRange), min(self.height, startY + gridRange + 1)):
-                    # If agent can see at least part of a cell, they should be allowed to consider it
-                    euclideanDistance = math.sqrt((deltaX - startX)**2 + (deltaY - startY)**2)
-                    if euclideanDistance < gridRange + 1 and self.grid[deltaX][deltaY] != self.grid[startX][startY]:
-                        cellsInRange.append({"cell": self.grid[deltaX][deltaY], "distance": euclideanDistance})
-        return cellsInRange
+            self.findCardinalCellRanges(maxDeltaX, maxDeltaY, cellCoords)
+
+    def findRadialCellRanges(self, maxDeltaX, maxDeltaY, maxDeltaRadius, cellCoords):
+        distanceTable = self.createDistanceTable(maxDeltaX, maxDeltaY)
+        numCells = self.width * self.height
+        for i in range(numCells):
+            x1, y1 = cellCoords[i]
+            for j in range(i + 1, numCells):
+                x2, y2 = cellCoords[j]
+                deltaX = self.findWraparoundDistance(x1 - x2, self.width)
+                deltaY = self.findWraparoundDistance(y1 - y2, self.height)
+                if deltaX > maxDeltaX or deltaY > maxDeltaY:
+                    continue
+                deltaPair = tuple(sorted((deltaX, deltaY)))
+                distance = distanceTable[deltaPair]
```

Comment:
There's a runtime error for this when setting vision and movement modes to `radial`, neighborhood mode to `moore`, and keeping the rest of the default config the same (exception: using the `bentham` model for survivability):

```
python sugarscape.py --conf config.json
Traceback (most recent call last):
  File "/home/user/sugarscape/sugarscape.py", line 1047, in <module>
    S = Sugarscape(configuration)
        ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/colinscape/sugarscape.py", line 42, in __init__
    self.configureEnvironment(configuration["environmentMaxSugar"], configuration["environmentMaxSpice"], configuration["environmentSugarPeaks"], configuration["environmentSpicePeaks"])
  File "/home/user/colinscape/sugarscape.py", line 200, in configureEnvironment
    self.environment.findCellRanges()
  File "/home/user/colinscape/environment.py", line 131, in findCellRanges
    self.findRadialCellRanges(maxDeltaX, maxDeltaY, maxDeltaRadius, cellCoords)
  File "/home/user/colinscape/environment.py", line 147, in findRadialCellRanges
    distance = distanceTable[deltaPair]
               ~~~~~~~~~~~~~^^^^^^^^^^^
KeyError: (0, 25)
```

Reply:
Should be fixed. Wraparound calculations should always use the environment borders even if the max deltas are smaller than the borders.